worlds/cv64/stages.py

```diff
@@ -0,0 +1,446 @@
+    stages around, there is one particularly big rule that must be kept in mind to ensure things don't get too wacky.
+    That being:
+
+    Villa and Castle Center cannot appear in branching path stage slots; they can only be on "main" path slots.
+
+    So for this reason, generating a new stage layout is not as simple as just scrambling a list of stages around. It
+    must be done in such a way that whatever stages directly follow Villa or CC is not the other stage. The exception is
+    if branching stages are not a thing at all due to the player settings, in which case everything I said above does
+    not matter. Consider the following representation of a stage "timeline", wherein each "-" represents a main stage
+    and a "=" represents a pair of branching stages:
+
+    -==---=---
+
+    In the above example, CC is the first "-" and Villa is the fourth. CC and Villa can only be "-"s whereas every other
+    stage can be literally anywhere, including on one of the "=" dashes. Villa will always be followed by one pair of
+    branching stages and CC will be followed by two pairs.
+
+    This code starts by first generating a singular list of stages that fit the criteria of Castle Center not being in
+    the next two entries following Villa and Villa not being in the next four entries after Castle Center. Once that has
+    been figured out, it will then generate a dictionary of stages with the appropriate information regarding what
+    stages come before and after them to then be used for Entrance creation as well as what position in the list they
+    are in for the purposes of the spoiler log and extended hint information.
+
+    I opted to use the Rondo of Blood "'" stage notation to represent Carrie stage slots specifically. If a main stage
+    with a backwards connection connects backwards into a pair of branching stages, it will be the non-"'" stage
+    (Reinhardt's) that it connects to. The Carrie stage slot cannot be accessed this way.
+
+    If anyone has any ideas or suggestions on how to improve this, I'd love to hear them! Because it's only going to get
+    uglier come Legacy of Darkness and Cornell's funny side route later on.
+    """
+
+    # Verify the starting stage is valid. If it isn't, or if the mystery setting is being used, pick a stage at random.
+    if starting_stage_value < len(vanilla_stage_order)-1 and vanilla_stage_order[starting_stage_value] not in \
+            stage_1_blacklist and verify_character_stage(world, vanilla_stage_order[starting_stage_value]):
+        starting_stage = vanilla_stage_order[starting_stage_value]
+    else:
+        possible_stages = []
+        for stage in vanilla_stage_order:
+            if stage in active_stage_exits and stage != rname.castle_keep:
+                possible_stages.append(stage)
+        starting_stage = possible_stages[world.random.randrange(0, len(possible_stages) - 1)]
```

Comment:
Is it intentional that this excludes the last element of `possible_stages`?

Reply:
Once upon a time, I probably had it set up such that it was picking a stage from the vanilla stage list. And since Castle Keep is at the end of said list and I did not want it getting picked as the starting stage, that was my way of ensuring it wouldn't get picked. But at some point, I probably changed it to be a list of possible stages that I construct based on what stages are actually in the seed and didn't think to remove that - 1.

Reply:
You're right, I can now! XD